### ISSUE TITLE
unit tests fail with mpich + petsc

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#1916]
When running with mpich + petsc I am encountering the error:
```
Fatal error in MPI_Add_error_class: Other MPI error, error stack:
MPI_Add_error_class(95): MPI_Add_error_class(errorclass=0x7ffff3fad958) failed
MPI_Add_error_class(74): No more user-defined error classes
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=671687183
:
system msg for write_line failure : Bad file descriptor
[Inferior 1 (process 31862) exited with code 017]
```

The error happens as the PETSc library is initialised the 128th time.

I thus assume that this is an issue in MPICH, which does not support more than 128 custom error classes.

Can we get a work around?

**Comment by dschwoerer** [comment]
MPICH has a hard limit of 128 custom error classes:
https://github.com/pmodels/mpich/issues/4341

It would probably by best to initiallise PETSc only once and keep that instance around, or at least limit the number to well below 128. Another option would be to also restart mpi as well as PETSc ...

**Comment by ZedThree** [comment]
Restarting MPI is disallowed by the standard unfortunately. I think we can just stick an instance of `PetscLib` in the test `main()`.

**Comment by ZedThree** [comment]
Fixed in #1922
